plutus-contract/src/Language/Plutus/Contract/Test.hs

```diff
@@ -52,7 +56,19 @@ import qualified Wallet.Emulator                       as EM
 
 import           Language.Plutus.Contract.Trace        as X
 
-type TracePredicate a = InitialDistribution -> Predicate (ContractTraceResult a)
+newtype PredF f a = PredF { unPredF :: a -> f Bool }
+
+instance Contravariant (PredF f) where
+    contramap f (PredF p) = PredF (p . f)
+
+instance Applicative f => Semigroup (PredF f a) where
+    l <> r = PredF $ \a -> (&&) <$> unPredF l a <*> unPredF r a
```

Comment:
I'm not sure about this since it's privileging the AND monoid for booleans...

Reply:
Do you actually use this?

Reply:
yes I actually use it a lot in the tests, ANDing the predicates together there is the default mode in my opinion

Reply:
Actually, you know what the better solution is? Lattice!

Reply:
But fine for now. I'm just going to bite the bullet and write a lattice class.

Reply:
> Actually, you know what the better solution is? Lattice!

As usual ;)